packages/bsky/src/services/moderation/index.ts

```diff
@@ -0,0 +1,455 @@
+    if ('did' in subject) {
+      // Allowing dids that may not exist: may have been deleted but needs to remain actionable.
+      subjectInfo = {
+        subjectType: 'com.atproto.admin.defs#repoRef',
+        subjectDid: subject.did,
+        subjectUri: null,
+        subjectCid: null,
+      }
+      if (subjectBlobCids?.length) {
+        throw new InvalidRequestError('Blobs do not apply to repo subjects')
+      }
+    } else {
+      // Allowing records/blobs that may not exist: may have been deleted but needs to remain actionable.
+      subjectInfo = {
+        subjectType: 'com.atproto.repo.strongRef',
+        subjectDid: subject.uri.host,
+        subjectUri: subject.uri.toString(),
+        subjectCid: subject.cid.toString(),
+      }
+    }
+
+    const subjectActions = await this.getCurrentActions(subject)
+    if (subjectActions.length) {
+      throw new InvalidRequestError(
+        `Subject already has an active action: #${subjectActions[0].id}`,
+        'SubjectHasAction',
+      )
+    }
+
+    const actionResult = await this.db.db
+      .insertInto('moderation_action')
+      .values({
+        action,
+        reason,
+        createdAt: createdAt.toISOString(),
+        createdBy,
+        createLabelVals,
+        negateLabelVals,
+        ...subjectInfo,
+      })
+      .returningAll()
+      .executeTakeFirstOrThrow()
+
+    if (subjectBlobCids?.length && !('did' in subject)) {
+      const blobActions = await this.getCurrentActions({
+        cids: subjectBlobCids,
+      })
+      if (blobActions.length) {
+        throw new InvalidRequestError(
+          `Blob already has an active action: #${blobActions[0].id}`,
+          'SubjectHasAction',
+        )
+      }
+
+      await this.db.db
+        .insertInto('moderation_action_subject_blob')
+        .values(
+          subjectBlobCids.map((cid) => ({
+            actionId: actionResult.id,
+            cid: cid.toString(),
+          })),
+        )
+        .execute()
+    }
+
+    return actionResult
+  }
+
+  async logReverseAction(info: {
+    id: number
+    reason: string
+    createdBy: string
+    createdAt?: Date
+  }): Promise<ModerationActionRow> {
+    const { id, createdBy, reason, createdAt = new Date() } = info
+
+    const result = await this.db.db
+      .updateTable('moderation_action')
+      .where('id', '=', id)
+      .set({
+        reversedAt: createdAt.toISOString(),
+        reversedBy: createdBy,
+        reversedReason: reason,
+      })
+      .returningAll()
+      .executeTakeFirst()
+
+    if (!result) {
+      throw new InvalidRequestError('Moderation action not found')
+    }
+
+    return result
+  }
+
+  async takedownRepo(info: { takedownId: number; did: string }) {
+    await this.db.db
+      .updateTable('actor')
+      .set({ takedownId: info.takedownId })
+      .where('did', '=', info.did)
+      .where('takedownId', 'is', null)
+      .executeTakeFirst()
+  }
+
+  async reverseTakedownRepo(info: { did: string }) {
```

Comment:
should we be invalidating all related blobs here?

Reply:
That's a good question.  To date blobs haven't been a part of repo takedowns, but it's possible to start allowing it with a little effort.  Here's where that currently gets stopped: https://github.com/bluesky-social/atproto/blob/f30887bc1c21936ff2b79832d4f80e04edfdb9e2/packages/pds/src/services/moderation/index.ts#L215-L217